extendedcopy.go

```diff
@@ -182,52 +181,64 @@ func findRoots(ctx context.Context, storage content.ReadOnlyGraphStorage, node o
 // For performance consideration, when using both FilterArtifactType and
 // FilterAnnotation, it's recommended to call FilterArtifactType first.
 func (opts *ExtendedCopyGraphOptions) FilterAnnotation(key string, regex *regexp.Regexp) {
+	filter := func(filtered []ocispec.Descriptor, predecessors []ocispec.Descriptor) []ocispec.Descriptor {
+		for _, p := range predecessors {
+			if value, ok := p.Annotations[key]; ok && (regex == nil || regex.MatchString(value)) {
+				filtered = append(filtered, p)
+			}
+		}
+		return filtered
+	}
+
 	fp := opts.FindPredecessors
 	opts.FindPredecessors = func(ctx context.Context, src content.ReadOnlyGraphStorage, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
 		var predecessors []ocispec.Descriptor
 		var err error
+
+		referrerFinder, supportReferrers := src.(registry.ReferrerFinder)
 		if fp == nil {
+			if supportReferrers {
+				// if src is a ReferrerFinder, use Referrers() for possible memory saving
+				if err := referrerFinder.Referrers(ctx, desc, "", func(referrers []ocispec.Descriptor) error {
+					// for each page of the results, filter the referrers
+					predecessors = filter(predecessors, referrers)
+					return nil
+				}); err != nil {
+					return nil, err
+				}
+				return predecessors, nil
+			}
 			predecessors, err = src.Predecessors(ctx, desc)
 		} else {
 			predecessors, err = fp(ctx, src, desc)
 		}
 		if err != nil {
 			return nil, err
 		}
-		var filtered []ocispec.Descriptor
-		for _, p := range predecessors {
-			if p.Annotations == nil {
-				switch p.MediaType {
-				case docker.MediaTypeManifest, ocispec.MediaTypeImageManifest,
-					docker.MediaTypeManifestList, ocispec.MediaTypeImageIndex,
-					artifactspec.MediaTypeArtifactManifest:
-					if err = func() error {
-						rc, err := src.Fetch(ctx, p)
+
+		if !supportReferrers {
```

Comment:
`supportReferrers == true` does not imply `predecessors` comes from the referrer API. They may come from user specified `opts.FindPredecessors`.

Reply:
updated.